src/strategies/offer_maker/market_making/Kandel/ExplicitKandel.sol

```diff
@@ -0,0 +1,153 @@
+  uint96[] _baseOfIndex;
+  ///@notice quote distribution: `quoteOfIndex[i]` is the amount of quote tokens Kandel must give or want at index i
+  uint96[] _quoteOfIndex;
+
+  constructor(IMangrove mgv, IERC20 base, IERC20 quote, uint gasreq, uint16 nslots)
+    CoreKandel(mgv, base, quote, gasreq, nslots)
+  {
+    _baseOfIndex = new uint96[](nslots);
+    _quoteOfIndex = new uint96[](nslots);
+  }
+
+  function __reserve__(address) internal view override returns (address) {
+    return address(this);
+  }
+
+  ///@notice sets the base/quote distribution that Kandel should follow using continuous slices
+  ///@param from start index (included). Must be less than `to`.
+  ///@param to end index (excluded). Must be less than `NSLOT`.
+  ///@param slice slice[i][0/1] is the distribution of base/quote at index i.
+  function setDistribution(uint from, uint to, uint[][2] calldata slice) external onlyAdmin {
+    for (uint i = from; i < to; i++) {
+      uint sliceIndex = i - from;
+      require(uint96(slice[0][sliceIndex]) == slice[0][sliceIndex], "Kandel/baseOverflow");
+      require(uint96(slice[1][sliceIndex]) == slice[1][sliceIndex], "Kandel/quoteOverflow");
+      _baseOfIndex[i] = uint96(slice[0][sliceIndex]);
+      _quoteOfIndex[i] = uint96(slice[1][sliceIndex]);
+    }
+  }
+
+  ///@inheritdoc AbstractKandel
+  function baseOfIndex(uint index) public view override returns (uint96) {
+    return _baseOfIndex[index];
+  }
+
+  ///@inheritdoc AbstractKandel
+  function quoteOfIndex(uint index) public view override returns (uint96) {
+    return _quoteOfIndex[index];
+  }
+
+  function baseDist() external view onlyAdmin returns (uint96[] memory) {
+    return _baseOfIndex;
+  }
+
+  function quoteDist() external view onlyAdmin returns (uint96[] memory) {
+    return _quoteOfIndex;
+  }
+
+  ///@notice checks whether offer whose logic is being executed is currently the best on Mangove
+  ///@param order the taker order that is being executed
+  ///@dev `isBest` => `order.offer` is the best bid/ask of this strat (but the converse is not true in general).
+  function _isBest(MgvLib.SingleOrder calldata order) internal view returns (bool) {
+    uint lookup = MGV.offers(order.outbound_tkn, order.inbound_tkn, order.offerId).prev();
+    return lookup == 0 || MGV.offers(order.outbound_tkn, order.inbound_tkn, lookup).gives() == 0;
+  }
+
+  ///@inheritdoc AbstractKandel
+  function _transportLogic(OrderType ba, MgvLib.SingleOrder calldata order)
+    internal
+    virtual
+    override
+    returns (OrderType dualBa, uint dualIndex, OfferArgs memory args)
+  {
+    uint index = indexOfOfferId(ba, order.offerId);
+
+    if (index == 0) {
+      emit AllAsks(MGV, BASE, QUOTE);
+    }
+    if (index == NSLOTS - 1) {
+      emit AllBids(MGV, BASE, QUOTE);
+    }
+    dualIndex = dual(ba, index);
+    dualBa = dual(ba);
+
+    (MgvStructs.OfferPacked dualOffer, MgvStructs.OfferDetailPacked dualOfferDetails) = getOffer(dualBa, dualIndex);
```

Comment:
will dualOffer always exist at this point?

Reply:
I think I now understand that no, it does not always exist - and this is intentional in that all things are just 0 which also works. It means we'll use 0 for pivotid, but what else can we do.
It seems we do a lot of work even though we will (often?) end up with a crossed book on partial fills - i.e. should we check whether dual offer is alive before doing this work?
The code `dualIndex = dual(ba, index)` is that encoding the transportLogic of moving the bid one point? Then I would expect that method to be overridable but maybe that is not necessary. Should dual be renamed dualIndex?
I think the logic around _transportLogic and reposting is the core of the complexity, and maybe it would help if we could go through it in a call.